common/scala/src/main/scala/org/apache/openwhisk/core/database/mongodb/MongoDBArtifactStore.scala

```diff
@@ -0,0 +1,661 @@
+  implicit system: ActorSystem,
+  val logging: Logging,
+  jsonFormat: RootJsonFormat[DocumentAbstraction],
+  val materializer: ActorMaterializer,
+  docReader: DocumentReader)
+    extends ArtifactStore[DocumentAbstraction]
+    with DocumentProvider
+    with DefaultJsonProtocol
+    with AttachmentSupport[DocumentAbstraction] {
+
+  import MongoDBArtifactStore._
+
+  protected[core] implicit val executionContext = system.dispatcher
+
+  private val mongodbScheme = "mongodb"
+  val attachmentScheme: String = attachmentStore.map(_.scheme).getOrElse(mongodbScheme)
+
+  private val database = client.getDatabase(dbName)
+  private val collection = getCollectionAndCreateIndexes
+  private val gridFSBucket = GridFSBucket(database, collName)
+
+  private val jsonWriteSettings = JsonWriterSettings.builder().outputMode(JsonMode.RELAXED).build
+
+  // MongoDB doesn't support using `$` as the first char of field name, so below two fields needs to be encoded first
+  private val fieldsNeedEncode = Seq("annotations", "parameters")
+
+  override protected[database] def put(d: DocumentAbstraction)(implicit transid: TransactionId): Future[DocInfo] = {
+    val asJson = d.toDocumentRecord
+
+    val id: String = asJson.fields.getOrElse("_id", JsString.empty).convertTo[String].trim
+    require(!id.isEmpty, "document id must be defined")
+
+    val (old_rev, rev) = revisionCalculate(asJson)
+    val docinfoStr = s"id: $id, rev: $rev"
+    val start =
+      transid.started(this, LoggingMarkers.DATABASE_SAVE, s"[PUT] '$collName' saving document: '$docinfoStr'")
+
+    val encodedData = encodeFields(fieldsNeedEncode, asJson)
+
+    val data = JsObject(
+      encodedData.fields + (_computed -> documentHandler.computedFields(asJson)) + ("_rev" -> rev.toJson))
+
+    val filters =
+      if (rev.startsWith("1-")) {
+        // for new document, we should get no matched document and insert new one
+        // if there is a matched document, that one with no _rev filed will be replaced
+        // if there is a document with the same id but has an _rev field, will return en E11000(conflict) error
+        Filters.and(Filters.eq("_id", id), Filters.not(Filters.exists("_rev")))
+      } else {
+        // for old document, we should find a matched document and replace it
+        // if no matched document find and try to insert new document, mongodb will return an E11000 error
+        Filters.and(Filters.eq("_id", id), Filters.eq("_rev", old_rev))
+      }
+
+    val f =
+      collection
+        .findOneAndReplace(
+          filters,
+          Document(data.compactPrint),
+          FindOneAndReplaceOptions().upsert(true).returnDocument(ReturnDocument.AFTER))
+        .toFuture()
+        .map { doc =>
+          transid.finished(this, start, s"[PUT] '$collName' completed document: '$docinfoStr', document: '$doc'")
+          DocInfo(DocId(id), DocRevision(rev))
+        }
+        .recover {
+          case t: MongoException if t.getCode == 11000 =>
```

Comment:
Would be great to add some comment to describe what 11000(conflict) stands for or suggesting to use a constant for this.

Reply:
ok